File: java/src/main/java/io/binac/leetcode/ValidNumber.java
package io.binac.leetcode;

/**
 * <p>A <strong>valid number</strong> can be split up into these components (in order):</p>
 *
 * <ol>
 * 	<li>A <strong>decimal number</strong> or an <strong>integer</strong>.</li>
 * 	<li>(Optional) An <code>'e'</code> or <code>'E'</code>, followed by an <strong>integer</strong>.</li>
 * </ol>
 *
 * <p>A <strong>decimal number</strong> can be split up into these components (in order):</p>
 *
 * <ol>
 * 	<li>(Optional) A sign character (either <code>'+'</code> or <code>'-'</code>).</li>
 * 	<li>One of the following formats:
 * 	<ol>
 * 		<li>One or more digits, followed by a dot <code>'.'</code>.</li>
 * 		<li>One or more digits, followed by a dot <code>'.'</code>, followed by one or more digits.</li>
 * 		<li>A dot <code>'.'</code>, followed by one or more digits.</li>
 * 	</ol>
 * 	</li>
 * </ol>
 *
 * <p>An <strong>integer</strong> can be split up into these components (in order):</p>
 *
 * <ol>
 * 	<li>(Optional) A sign character (either <code>'+'</code> or <code>'-'</code>).</li>
 * 	<li>One or more digits.</li>
 * </ol>
 *
 * <p>For example, all the following are valid numbers: <code>["2", "0089", "-0.1", "+3.14", "4.", "-.9", "2e10", "-90E3", "3e+7", "+6e-1", "53.5e93", "-123.456e789"]</code>, while the following are not valid numbers: <code>["abc", "1a", "1e", "e3", "99e2.5", "--6", "-+3", "95a54e53"]</code>.</p>
 *
 * <p>Given a string <code>s</code>, return <code>true</code><em> if </em><code>s</code><em> is a <strong>valid number</strong></em>.</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 *
 * <pre><strong>Input:</strong> s = "0"
 * <strong>Output:</strong> true
 * </pre>
 *
 * <p><strong>Example 2:</strong></p>
 *
 * <pre><strong>Input:</strong> s = "e"
 * <strong>Output:</strong> false
 * </pre>
 *
 * <p><strong>Example 3:</strong></p>
 *
 * <pre><strong>Input:</strong> s = "."
 * <strong>Output:</strong> false
 * </pre>
 *
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li><code>1 &lt;= s.length &lt;= 20</code></li>
 * 	<li><code>s</code> consists of only English letters (both uppercase and lowercase), digits (<code>0-9</code>), plus <code>'+'</code>, minus <code>'-'</code>, or dot <code>'.'</code>.</li>
 * </ul>
 */
public class ValidNumber {
    public static class Solution1 {
        public boolean isNumber(String s) {
            final int len = s.length();
            int i = 0;
            int point = 0, exp = 0, digit = 0;
            for (; i < len; ++i) {
                final char ch = s.charAt(i);
                if (ch == '+' || ch == '-') {
                    if (i > 0 && s.charAt(i - 1) != 'e' && s.charAt(i - 1) != 'E')
                        return false;
                } else if ('0' <= ch && ch <= '9') {
                    ++digit;
                } else if (ch == '.') {
                    if (point > 0 || exp > 0)
                        return false;
                    ++point;
                } else if (ch == 'e' || ch == 'E') {
                    if (exp > 0 || digit == 0)
                        return false;
                    ++exp;
                    digit = 0;
                } else {
                    break;
                }
            }
            return i == len && digit > 0;
        }
    }
}

File: python3/leetcodepy/valid_number.py
"""
A valid number can be split up into these components (in order):

1. A decimal number or an integer.
2. (Optional) An 'e' or 'E', followed by an integer.

A decimal number can be split up into these components (in order):

1. (Optional) A sign character (either '+' or '-').
2. One of the following formats:
    1. One or more digits, followed by a dot '.'.
    2. One or more digits, followed by a dot '.', followed by one or more digits.
    3. A dot '.', followed by one or more digits.

An integer can be split up into these components (in order):

1. (Optional) A sign character (either '+' or '-').
2. One or more digits.

For example, all the following are valid numbers: ["2", "0089", "-0.1", "+3.14", "4.", "-.9", "2e10", "-90E3", "3e+7", "+6e-1", "53.5e93", "-123.456e789"], while the following are not valid numbers: ["abc", "1a", "1e", "e3", "99e2.5", "--6", "-+3", "95a54e53"].

Given a string s, return true if s is a valid number.



Example 1:

Input: s = "0"
Output: true

Example 2:

Input: s = "e"
Output: false

Example 3:

Input: s = "."
Output: false


Constraints:

1 <= s.length <= 20
s consists of only English letters (both uppercase and lowercase), digits (0-9), plus '+', minus '-', or dot '.'.
"""


class Solution1:
    def isNumber(self, s: str) -> bool:
        point, exp, digit = False, False, False
        i = 0
        while i < len(s):
            if s[i] == '+' or s[i] == '-':
                if i != 0 and s[i - 1] != 'e' and s[i - 1] != 'E':
                    return False
                i += 1
            elif '0' <= s[i] <= '9':
                digit = True
                i += 1
            elif s[i] == '.':
                if point or exp:
                    return False
                point = True
                i += 1
            elif s[i] == 'e' or s[i] == 'E':
                if exp or digit == 0:
                    return False
                exp += 1
                digit = 0
                i += 1
            else:
                break
        return i == len(s) and digit
